### ISSUE TITLE
Help with Rebricked GameShark pro

### ISSUE THREAD

**Issue by FisherMoraes** [parasyte/n64rd#5]
My GS Pro just self flashed itself, like the one in this video: https://www.youtube.com/watch?v=ATrbGD4H5eM&lc=z22suvlrmpqyu5hcbacdp43a5oy0ajw4blu5co0j0dtw03c010c.1506012745868420
I've seem the recomendation to put sockets on the flash ROMs and flash it again.
Did it and worked as a charm! :-)
Them again, I tried to play Lamborghini American Challenge on the saved GS.
Got the same screen again, but as it showed up I quickly turned off the N64.
Switched the game and tried again. No luck, it seemed to had corrupted the flash.
Flashed it again and no deal :-(
Any idea if this GS can possibly be saved again?
Thanks in advance.

**Comment by parasyte** [comment]
Flashing the ROMs directly (pulling them off the PCB) is foolproof. There is a *small* chance that one or both of the ROMs are damaged, in which case you can just replace them. Replacements cost only a few cents, so it's no big deal.

**Comment by FisherMoraes** [comment]
My bad! :-( It was some bad soldering.
Unfortunatelly, due to my lack of experience when I removed one of the chips I heated pulled it, taking some pieces of the board with it :-(
The other chip I just warmed the back of the PCB with it upside down and it popped out nicely :-)
An interesting thing is that it bricked twice with that game...
It seems that the GS ROMs were rewritten with the first 2mbit of that game.
Any known schematics of the N64 GS?
I've traced the ROMs connections, would this be a good place to share it?
Thanks!

**Comment by parasyte** [comment]
I haven't seen any schematics for it. In the past I've just used a multimeter to trace pins by hand when I had a need.

If you have a doc you'd like to share, it might be best to create a pull request here. For example, if you put it into a `docs` directory. Then we can keep it shared to everyone for posterity. I also have a few very old text files laying around that provide some useful information. Stuff like how to interact with the LED, GS button, and LPT port from the N64-side.

Anyway, yeah, the corruption problems are two-fold, in my opinion; The hardware is poorly manufactured, and the software is poorly written. If the hardware had the correct pin pitch at the cartridge edge and used a 16-bit flash ROM, it would have been pretty stable. The software suffers from the hardware design because it has to write two flash ROMs in parallel without the ability to watch the status of both chips independently.

I've written about this bug before, but in short it ends up causing minor timing conflicts between the two chips to get out of sync while writing. The sync issue is possible during reads as well, but is less problematic because the N64 cartridge bus is generally slower than the access timing on either chip.

I feel the hardware is so flawed that it's difficult to justify using it for anything serious. There are at least two alternatives that I know of which I would recommend, and those are [64drive](http://64drive.retroactive.be/) and [EverDrive 64](https://krikzz.com/store/home/28-everdrive-64-v3.html). That said, the GameShark is still a cool piece of gaming history, and it is kind of fun to learn about the internals.

**Comment by FisherMoraes** [comment]
Tried a pull request.
Hopefully I've done it right.
Sure it's bad designed and I really would like to get one of these, but unfortunatelly the taxes here in my country (Brazil) are insane!!
Even a new GS is abour 300 bucks (about US$90)!!
That's why I tried so hard to fix this one.
I'll try to complete a bit more the schematics, but for now, my kids are having fun.
They just love the combination of old games + cheat devices!! :-)
Should I add a bit more of informatio on how the stuff bricked?
Basically it was a combination of a pirate game + an adaptor to put an original in the back + GS.

**Comment by parasyte** [comment]
Wow, that is steep! Best of luck taking care of that GameShark!

I just found the branch on your fork with the schematic, and I was able to create a PR for it (thanks for the contribution!) I'll make a few adjustments and add my own docs to this branch as well.

**Comment by FisherMoraes** [comment]
I added a bit more information I could collect.
The thing is already working, finicky as ever!
I'm looking forward to see the documentation you'll be sharing.
Many thanks :-)
